File: src/adapters/pi/ignore.ts
/**
 * pi adapter 私有 ignore / exclude 匹配（签名冻结，见 docs/m1-plan.md §1.5）。
 *
 * glob 规则（冻结，禁止膨胀）：
 *  - 字面量字符逐字匹配；
 *  - `*` 匹配任意非 `/` 字符序列（可以为空）；
 *  - 模式以 `/` 结尾 = 目录前缀匹配：`sessions/` 命中 `sessions/x` 与 `sessions/a/b`；
 *  - 不支持 `?`（按字面字符处理）与 `**`（等价于单个 `*`，不跨 `/`）。
 * 不引入 picomatch。
 */

/** 归一化：去掉前导 './' 与 '/'，保留尾部 '/'（目录前缀语义依赖它）。 */
function normalize(input: string): string {
  let out = input;
  while (out.startsWith('./')) out = out.slice(2);
  while (out.startsWith('/')) out = out.slice(1);
  return out;
}

function escapeLiteral(ch: string): string {
  return ch.replace(/[.*+?^${}()|[\]\\]/g, '\\$&');
}

/** 把单个 glob 模式编译成正则**主体**（不带锚点）。 */
function globBody(pattern: string): string {
  let body = '';
  for (const ch of pattern) {
    if (ch === '*') body += '[^/]*';
    else body += escapeLiteral(ch);
  }
  return body;
}

/**
 * 单个模式匹配（不含目录前缀语义，调用方需自行传入 `xxx/` 形态）。
 * 尾部 `/` 的模式走目录前缀：`^glob(prefix)(/.*)?$`。
 */
export function globMatch(candidate: string, pattern: string): boolean {
  const path = normalize(candidate);
  const pat = normalize(pattern);
  if (pat === '') return false;
  if (pat.endsWith('/')) {
    const prefix = pat.slice(0, -1);
    if (prefix === '') return false;
    return new RegExp(`^${globBody(prefix)}(?:/.*)?$`).test(path);
  }
  return new RegExp(`^${globBody(pat)}$`).test(path);
}

/**
 * adapter 级忽略判定：relPath 相对 adapter root，命中任一 pattern 即为 true。
 * relPath 为目录时同样适用（`sessions/` 命中 relPath `sessions`）。
 */
export function matchesIgnore(relPath: string, patterns: string[]): boolean {
  if (!patterns || patterns.length === 0) return false;
  return patterns.some((pattern) => globMatch(relPath, pattern));
}

File: internal/adapter/ignore.go
package adapter

import "strings"

// normalizeGlobPath applies the path normalization frozen by the TypeScript
// adapter.  Adapter paths are POSIX-style paths even when the process runs on
// a platform whose native separator is different.  Leading ./ and / are
// ignored; a trailing / is deliberately retained because it means "directory
// prefix".
func normalizeGlobPath(input string) string {
	out := input
	for strings.HasPrefix(out, "./") {
		out = strings.TrimPrefix(out, "./")
	}
	for strings.HasPrefix(out, "/") {
		out = strings.TrimPrefix(out, "/")
	}
	return out
}

// GlobMatch reports whether candidate matches pattern using the adapter glob
// grammar:
//
//   - all characters other than * are literals;
//   - * matches zero or more characters other than /;
//   - a trailing / turns the pattern into a directory-prefix match;
//   - ? and ** have no special meaning (two stars are simply two * tokens).
//
// This intentionally is not filepath.Match or a general purpose glob
// implementation.  In particular, a star never crosses a path separator.
func GlobMatch(candidate, pattern string) bool {
	path := normalizeGlobPath(candidate)
	pat := normalizeGlobPath(pattern)
	if pat == "" {
		return false
	}

	if strings.HasSuffix(pat, "/") {
		prefix := strings.TrimSuffix(pat, "/")
		if prefix == "" {
			return false
		}
		if globMatchExact(path, prefix) {
			return true
		}
		// The TypeScript implementation is equivalent to
		// ^body(?:/.*)?$.  Try every candidate slash as the optional
		// suffix boundary.  globMatchExact still prevents * from
		// consuming a slash in the prefix.
		for i := 0; i < len(path); i++ {
			if path[i] == '/' && globMatchExact(path[:i], prefix) {
				return true
			}
		}
		return false
	}

	return globMatchExact(path, pat)
}

// MatchesIgnore reports whether any pattern matches relPath.
func MatchesIgnore(relPath string, patterns []string) bool {
	for _, pattern := range patterns {
		if GlobMatch(relPath, pattern) {
			return true
		}
	}
	return false
}

// globMatchExact matches one complete path string.  The implementation is a
// small wildcard matcher rather than a regexp compiler, so literals can never
// accidentally acquire regexp meaning.  The only wildcard is * and it is
// constrained to a single path segment.
func globMatchExact(candidate, pattern string) bool {
	// A standard two-pointer wildcard matcher is sufficient here.  Since * is
	// not allowed to consume '/', each star can backtrack only within the
	// current segment.
	ci, pi := 0, 0
	star := -1
	starCandidate := -1

	for ci < len(candidate) {
		if pi < len(pattern) && pattern[pi] != '*' && pattern[pi] == candidate[ci] {
			ci++
			pi++
			continue
		}
		if pi < len(pattern) && pattern[pi] == '*' {
			star = pi
			starCandidate = ci
			pi++
			continue
		}
		if star >= 0 && starCandidate < len(candidate) && candidate[starCandidate] != '/' {
			starCandidate++
			ci = starCandidate
			pi = star + 1
			continue
		}
		return false
	}

	for pi < len(pattern) && pattern[pi] == '*' {
		pi++
	}
	return pi == len(pattern)
}

// IsBareGlob reports the four broad allowEscape spellings rejected by config
// validation in the W4 core worker.  The scan itself intentionally does not
// reject them: matching semantics belong here, while policy validation belongs
// to config.go.
func IsBareGlob(pattern string) bool {
	switch normalizeGlobPath(pattern) {
	case "*", "*/", "**", "**/":
		return true
	default:
		return false
	}
}

// Lower-case aliases keep the package convenient for tests in package adapter
// while the exported forms are used by the other adapter packages.
func globMatch(candidate, pattern string) bool { return GlobMatch(candidate, pattern) }
func matchesIgnore(relPath string, patterns []string) bool {
	return MatchesIgnore(relPath, patterns)
}
